src/SonarScanner.MSBuild.PreProcessor/Roslyn/RoslynAnalyzerProvider.cs

```diff
@@ -99,71 +96,83 @@ private IEnumerable<string> WriteAdditionalFiles(string language, IEnumerable<So
         return Enumerable.Empty<string>();
     }
 
-    private bool TryWriteSonarLintXmlFile(string language, IEnumerable<SonarRule> activeRules, out string sonarLintXmlPath)
+    private bool TryWriteSonarLintXmlFile(IEnumerable<SonarRule> activeRules, out string sonarLintXmlPath)
     {
-        var langDir = Path.Combine(this.teamBuildSettings.SonarConfigDirectory, language);
+        var langDir = Path.Combine(teamBuildSettings.SonarConfigDirectory, language);
         Directory.CreateDirectory(langDir);
 
         sonarLintXmlPath = Path.Combine(langDir, "SonarLint.xml");
         if (File.Exists(sonarLintXmlPath))
         {
-            this.logger.LogDebug(Resources.RAP_AdditionalFileAlreadyExists, language, sonarLintXmlPath);
+            logger.LogDebug(Resources.RAP_AdditionalFileAlreadyExists, language, sonarLintXmlPath);
             return false;
         }
 
-        var content = RoslynSonarLint.GenerateXml(activeRules, this.sonarProperties, language);
-        this.logger.LogDebug(Resources.RAP_WritingAdditionalFile, sonarLintXmlPath);
+        var content = RoslynSonarLint.GenerateXml(activeRules, sonarProperties, language);
+        logger.LogDebug(Resources.RAP_WritingAdditionalFile, sonarLintXmlPath);
         File.WriteAllText(sonarLintXmlPath, content);
         return true;
     }
 
-    private IEnumerable<AnalyzerPlugin> FetchAnalyzerPlugins(string language, IEnumerable<SonarRule> activeRules)
+    private IEnumerable<AnalyzerPlugin> FetchAnalyzerPlugins(IEnumerable<SonarRule> activeRules)
     {
-        var partialRepoKeys = ActiveRulesPartialRepoKeys(activeRules);
-        IList<Plugin> plugins = new List<Plugin>();
+        IList<Plugin> plugins = [];
+
+        AddPlugins(plugins, PartialRepoKeys(sonarProperties));
```

Comment:
This transition from list of properties to list of valid plugins is hard to follow, as the logic happens all over the place.

What about first transforming the list of properties to a dictionary of configuration objects (Plugins) first, and then using only the valid ones?